pkg/agent/controller/trafficcontrol/controller.go

```diff
@@ -0,0 +1,1009 @@
+	if err != nil {
+		return err
+	}
+
+	// Update the TrafficControl state.
+	tcState.targetPortName = targetPortName
+	tcState.returnPortName = returnPortName
+	if tcState.action != tc.Spec.Action {
+		tcState.action = tc.Spec.Action
+		tcUpdated = true
+	}
+	if tcState.direction != tc.Spec.Direction {
+		tcState.direction = tc.Spec.Direction
+		tcUpdated = true
+	}
+
+	// Get the list of Pods applying to the TrafficControl.
+	var podObjects []*v1.Pod
+	if podObjects, err = c.filterPods(&tc.Spec.AppliedTo); err != nil {
+		return err
+	}
+
+	// Reserve the set of Pods in TrafficControl state as stale, and reinitialize the set in TrafficControl state.
+	stalePods := tcState.pods
+	tcState.pods = sets.NewString()
+	// Reserve the set of OF ports in TrafficControl state as old, and reinitialize the set in TrafficControl state.
+	oldOfPorts := tcState.ofPorts
+	tcState.ofPorts = sets.NewInt32()
+	// Iterate the list of Pods applying to the TrafficControl.
+	for _, pod := range podObjects {
+		podNN := k8s.NamespacedName(pod.Namespace, pod.Name)
+		// Insert the Pod to the reinitialized set in TrafficControl state and remove the Pod from the stale set.
+		tcState.pods.Insert(podNN)
+		stalePods.Delete(podNN)
+
+		// If the TrafficControl is not the effective TrafficControl for the Pod, do nothing.
+		if !c.bindPodToTrafficControl(podNN, tcName) {
+			continue
+		}
+
+		// If the TrafficControl is the effective TrafficControl for the Pod, insert the port to the new set in
+		// TrafficControl state.
+		podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(pod.Name, pod.Namespace)
+		if len(podInterfaces) == 0 {
+			klog.InfoS("Interfaces of Pod not found", "Pod", klog.KObj(pod))
+			continue
+		}
+		tcState.ofPorts.Insert(podInterfaces[0].OFPort)
+	}
+
+	// If the target port, direction and action of the TrafficControl is updated, the mark flows should be reinstalled.
+	// If the newly generated ofPort set is different from the stale ofPort set, the flows should be reinstalled, the mark
+	// flows should be also reinstalled.
+	if tcUpdated || len(utilsets.SymmetricDifferenceInt32(tcState.ofPorts, oldOfPorts)) != 0 {
+		var ofPorts []uint32
+		for _, port := range tcState.ofPorts.List() {
+			ofPorts = append(ofPorts, uint32(port))
+		}
+		if err = c.ofClient.InstallTrafficControlMarkFlows(tc.Name, ofPorts, targetOFPort, tc.Spec.Direction, tc.Spec.Action); err != nil {
+			return err
+		}
+	}
+	if len(stalePods) != 0 {
+		// Resync the Pods applying to the TrafficControl to be deleted.
+		c.podsResync(stalePods, tcName)
+	}
+
+	return nil
+}
+
+func (c *Controller) uninstallTrafficControl(tcName string, tcState *trafficControlState) error {
+	// Uninstall the mark flows of the TrafficControl.
+	if err := c.ofClient.UninstallTrafficControlMarkFlows(tcName); err != nil {
+		return err
+	}
+
+	// Release the target port from the deleted TrafficControl.
+	if _, _, err := c.releaseTrafficControlPort(tcState.targetPortName, tcName); err != nil {
```

Comment:
targetPortName could be empty if previous sync didn't succeed

Reply:
Is this case? 

when the first sync of a TrafficControl gets failed when creating target port, before the second sync, the TrafficControl is deleted, then `tcState.targetPortName` is empty.

Reply:
yes